service/storage_proxy.cc

```diff
@@ -258,6 +259,31 @@ class shared_mutation : public mutation_holder {
     }
 };
 
+// shared mutation, but gets sent as a hint
+class hint_mutation : public shared_mutation {
+public:
+    explicit hint_mutation(frozen_mutation_and_schema&& fm_a_s) : shared_mutation(std::move(fm_a_s)){
+    }
+    explicit hint_mutation(const mutation& m) : shared_mutation(m) {
+    }
+    virtual bool store_hint(db::hints::manager& hm, gms::inet_address ep, tracing::trace_state_ptr tr_state) override {
+        return false;
+    }
```

Comment:
Why not exception? But I do not insist.

Reply:
I wouldn't like to risk it. I can see a possibility where `store_hint` could be called at shutdown (`drain_on_shutdown` -> `timeout_cb` -> `hint_to_dead_endpoints` -> `store_hint`).

Reply:
You are calling send_to_endpoint with disabled hints, this should not happen

Reply:
I agree with @gleb-cloudius if this one is hit it would definitely be a bug and we would want to know about this ASAP. You may keep it as 'return false' but then, please, add a very scary error message there.

Reply:
Ok. In the end I changed it into an exception throw.

Reply:
The message in the exception is confusing. It's not that you would try to store a hint for a second time but rather you would store a hint for a hint.

Reply:
Fixed